backend/czi_hosted/data_common/cache.py

```diff
@@ -0,0 +1,219 @@
+import threading
+import time
+from collections import Callable
+from typing import Optional
+
+import typing
+
+from backend.common.errors import DatasetAccessError
+from contextlib import contextmanager
+
+from backend.czi_hosted.data_common.rwlock import RWLock
+
+
+class CacheItem(object):
+    """This class provides a base class for caching information.  The first time information is accessed,
+    it is located (via the lambda function passed to .get) and cached.  Later accesses use the cached version.
+    It may also be deleted by the DataCacheManager to make room.  While data is actively being used
+    (during the lifetime of an api request), a reader lock is locked.
+    During that time, the data cannot be removed or updated."""
+
+    def __init__(self):
+        self.data_lock = RWLock()
+        self.data = None
+
+    def get(self, cache_key: str, create_data_function: typing.Optional[typing.Callable[[str], object]] = None,
+            create_data_args: object = {}):
+        self.data_lock.r_acquire()
+        if self.data:
+            return self.data
+        self.data_lock.r_release()
+        if create_data_function is None:
+            return None
+
+        self.data_lock.w_acquire()
+        # the data may have been loaded while waiting on the lock
+        if not self.data:
+            try:
+                self.data = create_data_function(cache_key, **create_data_args)
+            except Exception:
+                # necessary to hold the reader lock after an exception, since
+                # the release will occur when the context exits.
+                self.data_lock.w_demote()
+                raise
+
+        # demote the write lock to a read lock.
+        self.data_lock.w_demote()
+        return self.data
+
+    def release(self):
+        """Release the reader lock"""
+        self.data_lock.r_release()
+
+    def delete(self):
+        """Clear resources used by this cache item, if self.data has a cleanup method, call it, if not
+        delete the cached information"""
+        with self.data_lock.w_locked():
+            if self.data:
+                try:
+                    self.data.cleanup()
+                except AttributeError:
+                    pass
+                self.data = None
+
+    def attempt_delete(self):
+        """Delete, but only if the write lock can be immediately locked.  Return True if the delete happened"""
+        if self.data_lock.w_acquire_non_blocking():
+            if self.data:
```

Comment:
Just an edge case here.. if self.data is None, then this function would return None. Is that expected/handled?

Reply:
Based on where it is called that should be fine, but to be safe I'll release the lock and return True (to ensure the cache entry is deleted) when self.data is None